AK/IpAddressCidr.h

```diff
@@ -0,0 +1,269 @@
+class IPAddressCidr {
+public:
+    enum class IPAddressCidrError {
+        CidrTooLong,
+        StringParsingFailed,
+    };
+
+    using IPAddress = Details::AddressTraits<AddressFamily>::IPAddress;
+
+    static constexpr ErrorOr<AddressFamily, IPAddressCidrError> create(IPAddress address, u8 length)
+    {
+        if (length > AddressFamily::MAX_LENGTH)
+            return IPAddressCidrError::CidrTooLong;
+
+        return AddressFamily(address, length);
+    }
+
+    constexpr IPAddress const& ip_address() const& { return m_address; }
+    constexpr u32 length() const { return m_length; }
+
+    constexpr void set_ip_address(IPAddress address) { m_address = address; }
+    constexpr ErrorOr<void, IPAddressCidrError> set_length(u32 length)
+    {
+        if (length > AddressFamily::MAX_LENGTH)
+            return IPAddressCidrError::CidrTooLong;
+
+        m_length = length;
+        return {};
+    }
+
+    constexpr static ErrorOr<AddressFamily, IPAddressCidrError> from_string(StringView string)
+    {
+        Vector<StringView> const parts = string.split_view('/');
+
+        if (parts.size() != 2)
+            return IPAddressCidrError::StringParsingFailed;
+
+        auto ip_address = IPAddress::from_string(parts[0]);
+        if (!ip_address.has_value())
+            return IPAddressCidrError::StringParsingFailed;
+
+        auto length = parts[1].to_number<u8>();
+        if (!length.has_value())
+            return IPAddressCidrError::StringParsingFailed;
+
+        return IPAddressCidr::create(ip_address.release_value(), length.release_value());
+    }
+
+#ifdef KERNEL
+    ErrorOr<NonnullOwnPtr<Kernel::KString>> to_string() const
+    {
+        return Kernel::KString::formatted("{}/{}", m_address, m_length);
+    }
+#else
+    ErrorOr<String> to_string() const
+    {
+        return String::formatted("{}/{}", m_address, m_length);
+    }
+#endif
+
+    constexpr bool operator==(IPAddressCidr const& other) const = default;
+    constexpr bool operator!=(IPAddressCidr const& other) const = default;
+
+protected:
+    constexpr IPAddressCidr(IPAddress address, u8 length)
+        : m_address(address)
+        , m_length(length)
+    {
+    }
+
+private:
+    IPAddress m_address;
+    u8 m_length;
+};
+
+template<>
+class AddressTraits<IPv4AddressCidr> {
+public:
+    using IPAddress = IPv4Address;
+};
+
+template<>
+class AddressTraits<IPv6AddressCidr> {
+public:
+    using IPAddress = IPv6Address;
+};
+
+}
+
+class IPv4AddressCidr : public Details::IPAddressCidr<IPv4AddressCidr> {
+public:
+    constexpr static u8 MAX_LENGTH = 32;
+
+    constexpr IPv4AddressCidr(IPv4Address address, u8 length)
+        : IPAddressCidr(address, length)
+    {
+    }
+
+    constexpr IPv4Address netmask() const
+    {
+        IPv4Address netmask;
+        u8 free_bits = MAX_LENGTH - length();
+
+        if (free_bits == 32) {
+            netmask = IPv4Address(0, 0, 0, 0);
+        } else {
```

Comment:
Just asking because I'm not too knowledgeable about this, are we special casing this because we would otherwise be shifting a 32-bit wide type by 32 bits? I think I remember a discussion about that, not sure if the end result was "this is UB" or not.

Reply:
Shifting by the lengths of the type is UB as far as I know and tests related to having a length of 32/128 are sometimes failing without this check.